### ISSUE TITLE
Investigate node shutdown exception during network partitioning 

### ISSUE THREAD

**Issue by jumaffre** [microsoft/CCF#2571]
Testing the partitioning of nodes in #2553 has revealed another libuv-related exception on shutdown (see [CI failure](https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=25011&view=logs&j=ee9484ec-b476-577b-aa0e-d5ddfcbcd149&t=8e7d12cf-69d2-5704-84b7-b6fd04033d7b&l=207)):
```
63: 11:05:28.864 | ERROR    | infra.remote:log_errors:86 - Contents of /mnt/build/2/s/build/workspace/partitions_cft_2/err:
63: Stack trace (most recent call last):
63: #23   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
63: #22   Object "./cchost.virtual", at 0x408719, in 
63: #21   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fb9d6a66bf6, in __libc_start_main
63: #20   Object "./cchost.virtual", at 0x423189, in 
63: #19   Object "/usr/lib/x86_64-linux-gnu/libuv.so.1", at 0x7fb9d7ce5ddd, in uv_run
63: #18   Object "/usr/lib/x86_64-linux-gnu/libuv.so.1", at 0x7fb9d7cf05db, in uv__stream_destroy
63: #17   Object "./cchost.virtual", at 0x52cdf7, in 
63: #16   Object "./cchost.virtual", at 0x52ce51, in 
63: #15   Object "./cchost.virtual", at 0x52c29a, in 
63: #14   Object "./cchost.virtual", at 0x5c1974, in 
63: #13   Object "./cchost.virtual", at 0x5c1b0c, in 
63: #12   Object "./cchost.virtual", at 0x5c1b4e, in 
63: #11   Object "./cchost.virtual", at 0x5c1b7c, in 
63: #10   Object "./cchost.virtual", at 0x5c1bd3, in 
63: #9    Object "./cchost.virtual", at 0x5c1c8c, in 
63: #8    Object "./cchost.virtual", at 0x5c1dc9, in 
63: #7    Object "./cchost.virtual", at 0x5c2250, in 
63: #6    Object "./cchost.virtual", at 0x5c230c, in 
63: #5    Object "./cchost.virtual", at 0x4b3e4c, in 
63: #4    Object "./cchost.virtual", at 0x4b3e83, in 
63: #3    Object "./cchost.virtual", at 0x4b3fb8, in 
63: #2    Object "./cchost.virtual", at 0x439ca4, in 
63: #1    Object "./cchost.virtual", at 0x551ef8, in 
63: #0    Object "./cchost.virtual", at 0x44d825, in 
63: Segmentation fault (Address not mapped to object [0x1f])
```

I believe this can be circumvented by cleaning up the iptables rules before shutting down the nodes on network shutdown, but should be investigated further.

**Comment by jumaffre** [comment]
I believe the issue is that when we manually run the uv loop to avoid memory leaks when a node shuts down, some handles are reconnecting (even more likely for the partitioning test). At this point, the connect callbacks are run by libuv (see https://github.com/libuv/libuv/blob/v1.18.0/src/unix/stream.c#L448), which causes reconnection, which appends messages to the `reconnect_queue` in `node_connections.h`. The issue is that object has actually been destroyed at this point (to avoid it reading new messages off the ring buffer), hence the segfault. Working on a fix now that I'll PR shortly.

**Comment by jumaffre** [comment]
Fixed by #2618
